use smb volume for case insensitive jobs

brettfo opened the pull request:
## Problem

More mature .NET repos with NuGet packages are commonly only built on Windows with a case-insensitive filesystem.  This can cause issues with the dependabot updater if there is a mismatch between the files in the repo and their casing in import statements.  Consider a `.csproj` with a line like `<Import Project="build/versions.props" />` but in the repo and on disk the file is named `build/Versions.props` (notice the upper-case `V`).  Since dependabot runs in a Linux host, all operations will fail because the requested files could not be loaded.  This same issue can arise if a NuGet package consumed by a repo has package-internal `<Import>` elements that don't match the casing of the file in the package.  All of these scenarios have been encountered.

## The Fix

To address this, dependabot needs to run on a case-insensitive filesystem, but that would add certain requirements for the Docker host that's running the update job.  To avoid this, we can use some functionality built into Docker itself, namely its ability to natively mount SMB/CIFS shares for use as volumes to bind to another container.  The SMB/CIFS sharing protocol was picked because it has an optional `nocase` mount option that makes all operations case-insensitive.

To accomplish this, the Docker host needs access to a SMB/CIFS share, but to again not impose additional requirements on the Docker host, a separate image is used, `ghcr.io.dependabot/dependabot-storage`.  That container only hosts a SMB/CIFS share on the default ports with a hardcoded username and password.  The default Docker configuration doesn't allow guest access so user/pass had to be used isntead.

## The changes to this codebase

Immediately before the updater container is created (e.g., from the `dependabot-updater-nuget` image), a new container is created from the `dependabot-storage` image and it is given no external internet access.  The host (this code) then mounts the SMB/CIFS share twice.  The first mount is with the standard options which is then given to the updater in the environment variable `$DEPENDABOT_REPO_CONTENTS_PATH`.  The second mount is exactly the same as the first, except with the `nocase` option and it is exposed to the updater in the environment variable `$DEPENDABOT_CASE_INSENSITIVE_REPO_CONTENTS_PATH`.

The job the proceeds as normal and the container cleanup operations have been amended to also remove the storage container and both volume mounts.

All of this work is behind a new feature flag `use_case_insensitive_filesystem`.

This experiment is not expected to be enabled for non-NuGet jobs, but I've done some testing anyway and found no change in behavior because the updater reads its version of `$DEPENDABOT_REPO_CONTENTS_PATH` and proceeds as normal.  Nothing is done with the `..._CASE_INSENSITIVE_...` variable.

The changes here will also require a change to `dependabot/dependabot-core` (PR pending) where the updater is aware of both directory paths.  It performs all update work against the case-insensitive one then normalizes all file paths (e.g. the contents of `create_pull_request`) using the case-sensitive one.

jakecoffman commented:
Is there any way we can test this? For instance making [one of these script tests](https://github.com/dependabot/cli/blob/main/testdata/scripts/basic.txt) we could test it's possible to mount a case-insensitive filesystem, but more importantly ensure it doesn't break going forward.

schmittjoseph commented on internal/infra/updater.go:
```diff
@@ -113,6 +141,127 @@ func NewUpdater(ctx context.Context, cli *client.Client, net *Networks, params *
+
+	// start storage container
+	if err = cli.ContainerStart(ctx, storageContainer.ID, container.StartOptions{}); err != nil {
+		tryRemoveStorageVolume(cli, ctx, caseSensitiveVolumeName)
+		tryRemoveStorageVolume(cli, ctx, caseInsensitiveVolumeName)
```
Consider deferring an anonymous function to perform the storage volume cleanup when exiting with a non-nill `err`. That way we don't need to remember to do this cleanup on all of the potential error paths.

brettfo commented:
@jakecoffman I added `smb-mount.txt` that writes a file to the case-sensitive share then reads it back with different casing from the case insensitive share.  Good catch on this, turns out I also forgot to add the code to pull the storage image, but this caught it.  "Works on my machine!"

@schmittjoseph Addressed feedback in additional commit.

Current smoke failures are due to a rate limit, I'll see what I can do to get those cleared.

kbukum1 reviewed (approved):
LGTM
